src/Gateway/Connection.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Ragnarok\Fenrir\Gateway;
+
+use Evenement\EventEmitter;
+use Ragnarok\Fenrir\Bitwise\Bitwise;
+use Ragnarok\Fenrir\Constants\Events as Events;
+use Ragnarok\Fenrir\Constants\WebsocketEvents;
+use Ragnarok\Fenrir\Gateway\Objects\Payload;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Ragnarok\Fenrir\DataMapper;
+use Ragnarok\Fenrir\EventHandler;
+use Ragnarok\Fenrir\Gateway\Events\Ready;
+use Ragnarok\Fenrir\Websocket;
+use Ratchet\RFC6455\Messaging\MessageInterface;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+
+class Connection
+{
+    public const DEFAULT_WEBSOCKET_URL = 'wss://gateway.discord.gg/?v=10';
+
+    public EventHandler $events;
+
+    public Websocket $websocket;
+
+    public EventEmitter $raw;
+
+    private ?TimerInterface $heartbeatTimer;
+
+    private ?int $sequence = null;
+
+    private ?TimerInterface $scheduledReconnect;
+    private string $sessionId;
+    private string $reconnectUrl;
+
+    private Puppet $puppet;
+
+    /**
+     * @param Bitwise<\Ragnarok\Fenrir\Enums\Gateway\Intents> $intents
+     */
+    public function __construct(
+        private LoopInterface $loop,
+        private string $token,
+        private Bitwise $intents,
+        private DataMapper $mapper,
+        private LoggerInterface $logger = new NullLogger(),
+        int $timeout = 10
+    ) {
+        $this->raw = new EventEmitter();
+        $this->events = new EventHandler($this->mapper);
+
+        $this->websocket = new Websocket($timeout, $this->logger);
+        $this->puppet = new Puppet($this->websocket, $this->logger);
+
+        $this->websocket->on(WebsocketEvents::MESSAGE, function (MessageInterface $message) {
+            /** @var Payload */
+            $payload = $this->mapper->map(json_decode((string) $message), Payload::class);
+
+            $this->raw->emit($payload->op, [$payload]);
+        });
+
+        $this->setListeners();
+    }
+
+    private function setListeners(): void
+    {
+        $this->raw->on(0, function (Payload $payload) {
+            if (isset($payload->s) && $payload->s > $this->sequence) {
+                $this->sequence = $payload->s;
+            }
+
+            $this->events->handle($payload);
+        });
```

Comment:
```suggestion
    private function setListeners(): void
    {
       if ($this->listenersSet) {
           return;
       }
    
        $this->raw->on(0, function (Payload $payload) {
            if (isset($payload->s) && $payload->s > $this->sequence) {
                $this->sequence = $payload->s;
            }

            $this->events->handle($payload);
        });
```
*you'd of course have to add `private bool $listenersSet = false;`*